File: backend/app/agents/report_writer.py
"""Report Writer — aggregates pipeline results into a structured summary."""

from __future__ import annotations

import re
from collections import Counter
from typing import Any

from app.agents.base_agent import BaseAgent
from app.graph.state import FinalReport

_SUMMARY_PROMPT = """\
You are a senior legal analyst. Write an executive summary for a contract review.
We are writing this review from the perspective of the {represented_party}.

Contract: {contract_name}
Total clauses: {clause_count}
High-risk clauses (Severity HIGH or CRITICAL): {high_risk_count}
List of High-risk clauses (strictly derived programmatically):
{high_risk_clauses_list}

Compliance violation clauses: {compliance_violation_clauses}
Compliance violation issues (total): {compliance_violation_issues}
Overall risk score (0–100): {overall_score}

Top risk flags across all clauses:
{top_flags}

Top compliance issues:
{top_compliance}

Write a professional executive summary (3–5 paragraphs) suitable for a general
counsel to read in under 2 minutes. Follow these guidelines strictly:
1. Overall risk posture: Summarize the high-level legal risk of the contract from the {represented_party}'s perspective.
2. Most critical issues (by name): Focus ONLY on high/medium risk items. Do NOT mention low-risk or favorable terms (such as "favorable warranty" or "standard representations") as key risks.
3. Compliance violations & Systemic Gaps: Consolidate repeated compliance gaps (e.g. lack of data protection/DPA provisions flagged across multiple clauses) into one systemic issue, rather than listing them as separate, independent violations.
4. Actionable Next Steps & Numeric Grounding: Provide concrete, specific recommendations. If you recommend specific cure periods, liability caps, or other numeric remediation terms, you MUST either:
   - Ground them directly in the existing values present in the contract (from the Grounding Reference Points below).
   - If no reference value exists in the contract, you MUST explicitly mark them as illustrative placeholders in your output (e.g. '30 days (illustrative placeholder)' or '$100k (illustrative placeholder)'). Never present an invented number as a calculated recommendation without this placeholder designation.

Grounding Reference Points:
{grounding_anchors}

Plain text only — no JSON, no markdown headers. Do NOT include any conversational preamble or introduction.
""".strip()


def clean_summary_text(summary: str) -> str:
    # 1. Remove markdown code blocks (e.g. ```json ... ``` or ``` ... ```)
    summary = re.sub(r"```[\s\S]*?```", "", summary)

    # 2. Filter out raw JSON lines and conversational introductions
    lines = summary.splitlines()
    cleaned_lines = []
    in_json = False
    for line in lines:
        stripped = line.strip()
        if stripped.startswith("{") or stripped.startswith("["):
            in_json = True
            continue
        if in_json:
            if stripped.endswith("}") or stripped.endswith("]"):
                in_json = False
            continue

        # Remove standard intro preambles.
        # FIX: this list previously required an exact substring match for
        # each phrasing variant (e.g. it had "here is a rewritten" but not
        # the contraction "here's a rewritten", which is exactly the phrase
        # that leaked through in production: "Here's a rewritten executive
        # summary that corrects the fabricated quote and removes it:").
        # The regex step below is the real backstop for this, but this list
        # is kept in sync with it as a fast first pass.
        lower_line = stripped.lower()
        if any(prefix in lower_line for prefix in [
            "here is a rewritten",
            "here's a rewritten",
            "here is the summary",
            "here is the executive summary",
            "here's the summary",
            "here's the executive summary",
            "here is a summary",
            "here is the json representation",
            "here's a summary",
            "here's the json representation",
            "rewritten executive summary",
        ]):
            continue

        cleaned_lines.append(line)

    summary = "\n".join(cleaned_lines).strip()

    # Remove any leading conversational phrases ending with a colon.
    # FIX: broadened alternation to explicitly include "here's a rewritten"
    # and "here's the ..." variants so this doesn't depend solely on the
    # line-level substring list above catching every contraction.
    summary = re.sub(
        r"^(here is|here's|this is|please find|summary of|executive summary of|"
        r"rewritten executive summary|here's a rewritten|here is a rewritten)"
        r"[\s\S]*?:\s*",
        "",
        summary,
        flags=re.IGNORECASE,
    )

    return summary.strip()


# FIX: previously only two specific JSON key names ("executive_summary",
# "overall_score") were checked, which happened to not match any of the
# actual keys observed in real leaked payloads ("risk_score",
# "key_clause_risks", "high_risk_clauses", "recommended_changes",
# "liability_cap", etc.). A generic structural check catches any
# `"key": value` pattern regardless of what the key is named, rather than
# maintaining an ever-growing hardcoded list that will always be one step
# behind whatever key names the model happens to produce next.
_JSON_KEY_VALUE_PATTERN = re.compile(r'"\w+"\s*:\s*(?:"|\[|\{|-?\d)')


def _has_ungrounded_dollar_or_day_figures(summary: str, anchors_text: str) -> bool:
    dollar_pattern = re.compile(r"\$\s?[\d,]+(?:\.\d+)?")
    day_pattern = re.compile(r"\b\d+\s?-?days?\b|\b\w+\s?\(\d+\)\s?-?days?\b", re.IGNORECASE)

    dollars = dollar_pattern.findall(summary)
    days = day_pattern.findall(summary)

    anchors_lower = anchors_text.lower()
    summary_lower = summary.lower()

    for dollar in dollars:
        dollar_norm = dollar.lower().strip()
        if dollar_norm not in anchors_lower:
            pos = summary_lower.find(dollar_norm)
            while pos != -1:
                context = summary_lower[pos:pos+60]
                if "illustrative placeholder" not in context:
                    return True
                pos = summary_lower.find(dollar_norm, pos + 1)

    for day in days:
        day_norm = day.lower().strip()
        digits = re.findall(r"\d+", day_norm)
        if digits:
            digit = digits[0]
            if digit not in anchors_lower:
                pos = summary_lower.find(day_norm)
                while pos != -1:
                    context = summary_lower[pos:pos+60]
                    if "illustrative placeholder" not in context:
                        return True
                    pos = summary_lower.find(day_norm, pos + 1)

    return False


def validate_summary(summary: str, anchors_text: str = "") -> bool:
    if not summary:
        return False
    if len(summary) < 200:
        return False
    if summary.startswith("{") or summary.startswith("["):
        return False
    if "```" in summary:
        return False
    # Generic JSON-shape detector instead of a hardcoded key list.
    if len(_JSON_KEY_VALUE_PATTERN.findall(summary)) >= 2:
        return False
    # Catch any residual meta-commentary about the generation process itself
    # that the line/regex cleanup above might have missed.
    lower = summary.lower()
    if any(
        phrase in lower
        for phrase in [
            "here is the json representation",
            "here's the json representation",
            "rewritten executive summary",
            "as an ai",
            "as a language model",
        ]
    ):
        return False

    if anchors_text and _has_ungrounded_dollar_or_day_figures(summary, anchors_text):
        return False

    return True


class ReportWriter(BaseAgent):
    agent_name = "report_writer"
    max_tokens = 2048

    def _execute(self, state: dict[str, Any]) -> dict[str, Any]:
        scores = state.get("risk_scores", [])
        compliance = state.get("compliance_results", [])
        edits = state.get("redline_edits", [])
        clauses = state.get("clauses", [])
        represented_party = state.get("represented_party", "Client")

        # Top 3 highest-risk average calculation
        if scores:
            sorted_scores = sorted([s.score for s in scores], reverse=True)
            top_3_scores = sorted_scores[:3]
            overall_score = round(sum(top_3_scores) / len(top_3_scores))
        else:
            overall_score = 0

        # Programmatic high-risk metrics (severity "high" or "critical")
        high_severity_clauses = [s for s in scores if s.severity in ("high", "critical")]
        high_risk_count = len(high_severity_clauses)

        clause_map = {c.id: c for c in clauses}
        high_risk_clauses_info = []
        for s in high_severity_clauses:
            c = clause_map.get(s.clause_id)
            heading = c.heading if c else "Unknown Heading"
            high_risk_clauses_info.append(f"- Clause {s.clause_id} ({heading}): Severity={s.severity.upper()}, Score={s.score}")
        high_risk_clauses_text = "\n".join(high_risk_clauses_info) if high_risk_clauses_info else "None"

        # Explicitly distinguish compliance violation counts (clauses vs. issues)
        compliance_violation_clauses = state.get("compliance_violation_clauses", sum(1 for r in compliance if len(r.violations) > 0))
        compliance_violation_issues = state.get("compliance_violation_issues", sum(len(r.violations) for r in compliance))

        # Extract potential grounding anchors from clauses text to prevent hallucinated numbers
        anchors = []
        for c in clauses:
            text_lower = c.text.lower()
            if "limit" in text_lower or "cap" in text_lower:
                money_finds = re.findall(r"\$\s?[\d,]+(?:\.\d+)?", c.text)
                if money_finds:
                    anchors.append(f"Clause {c.id} ('{c.heading}') mentions liability limit/cap: {', '.join(money_finds)}")
            if "cure" in text_lower or "remedy" in text_lower or "terminate" in text_lower:
                days_finds = re.findall(r"\b\d+\s?days?\b|\b\w+\s?\(\d+\)\s?days?\b", c.text, re.IGNORECASE)
                if days_finds:
                    anchors.append(f"Clause {c.id} ('{c.heading}') mentions notice/cure period: {', '.join(days_finds)}")
        anchors_text = "\n".join(f"- {a}" for a in anchors) if anchors else "None found in source text."

        # Aggregate flags (exclude low risk/favorable flags to avoid executive summary mismatch)
        all_flags: list[str] = []
        for s in scores:
            if s.score >= 30:
                all_flags.extend(s.flags)
        
        top_flags = [f for f, _ in Counter(all_flags).most_common(5)]

        top_compliance: list[str] = []
        for c in compliance:
            if not c.compliant:
                for v in c.violations:
                    top_compliance.append(f"{v.framework}: {v.description}")
        
        # De-duplicate identical compliance entries to avoid cluttering the summary
        unique_compliance = []
        seen = set()
        for tc in top_compliance:
            if tc not in seen:
                seen.add(tc)
                unique_compliance.append(tc)
        top_compliance = unique_compliance[:5]

        summary_prompt = _SUMMARY_PROMPT.format(
            represented_party=represented_party,
            contract_name=state.get("contract_name", "Unknown"),
            clause_count=len(clauses),
            high_risk_count=high_risk_count,
            high_risk_clauses_list=high_risk_clauses_text,
            compliance_violation_clauses=compliance_violation_clauses,
            compliance_violation_issues=compliance_violation_issues,
            overall_score=overall_score,
            top_flags="\n".join(f"- {f}" for f in top_flags) or "None",
            top_compliance="\n".join(f"- {c}" for c in top_compliance) or "None",
            grounding_anchors=anchors_text,
        )

        executive_summary = ""
        for attempt in range(1, 4):
            raw_summary = self._call_llm(summary_prompt)
            cleaned = clean_summary_text(raw_summary)
            if validate_summary(cleaned, anchors_text):
                executive_summary = cleaned
                break
            else:
                self.log.warning("summary_validation_failed", attempt=attempt, raw_length=len(raw_summary))
                # Append critical instruction for retry
                summary_prompt = (
                    _SUMMARY_PROMPT.format(
                        represented_party=represented_party,
                        contract_name=state.get("contract_name", "Unknown"),
                        clause_count=len(clauses),
                        high_risk_count=high_risk_count,
                        high_risk_clauses_list=high_risk_clauses_text,
                        compliance_violation_clauses=compliance_violation_clauses,
                        compliance_violation_issues=compliance_violation_issues,
                        overall_score=overall_score,
                        top_flags="\n".join(f"- {f}" for f in top_flags) or "None",
                        top_compliance="\n".join(f"- {c}" for c in top_compliance) or "None",
                        grounding_anchors=anchors_text,
                    )
                    + "\n\nCRITICAL: Your previous response contained conversational comments, raw JSON, markdown code blocks, or was too short. DO NOT output JSON or any conversational preambles like 'Here is the summary'. Write ONLY the clean, user-facing executive summary paragraphs directly. Plain text only."
                )

        if not executive_summary:
            executive_summary = cleaned if cleaned else "Failed to generate executive summary."

        # Prepend perspective statement:
        prefix = f"This analysis is prepared from the {represented_party}'s perspective.\n\n"
        if not executive_summary.startswith("This analysis is prepared"):
            executive_summary = prefix + executive_summary

        # FIX (document-integrity, silent clause deletion): if clause_extractor
        # detected that fewer clauses were extracted than the source document's
        # headings suggest, that warning must not be dropped silently on the
        # floor — it gets surfaced as a first-class, visible field so the UI
        # can render a blocking banner (mirroring the pattern already used for
        # the redline/mitigation-score-mismatch signal).
        extraction_warning = state.get("extraction_integrity_warning")
        if extraction_warning:
            self.log.error("report_includes_extraction_integrity_warning", warning=extraction_warning)

        report = FinalReport(
            contract_id=state["contract_id"],
            contract_name=state.get("contract_name", "Unknown"),
            overall_score=overall_score,
            clause_count=len(clauses),
            high_risk_count=high_risk_count,
            violation_count=compliance_violation_issues,
            compliance_violation_clauses=compliance_violation_clauses,
            compliance_violation_issues=compliance_violation_issues,
            executive_summary=executive_summary,
            risk_scores=scores,
            compliance_results=compliance,
            redline_edits=edits,
            represented_party=represented_party,
            extraction_integrity_warning=extraction_warning,
        )

        self.log.info(
            "report_written",
            contract_id=state["contract_id"],
            overall_score=overall_score,
            represented_party=represented_party,
        )
        return {"report": report, "pipeline_complete": True}

File: pipeline-worker/src/agents/report-writer.ts
import { BaseAgent } from './base-agent';
import { FinalReport, ContractReviewState } from '../pipeline/state';
import { SUMMARY_PROMPT } from '../prompts/summary';

const JSON_KEY_VALUE_PATTERN = /"\w+"\s*:\s*(?:"|\[|\{|-?\d)/g;

export function cleanSummaryText(summary: string): string {
  // 1. Remove markdown code blocks (e.g. \`\`\`json ... \`\`\` or \`\`\` ... \`\`\`)
  summary = summary.replace(/```[\s\S]*?```/g, '');

  // 2. Filter out raw JSON lines and conversational introductions
  const lines = summary.split(/\r?\n/);
  const cleanedLines: string[] = [];
  let inJson = false;

  for (const line of lines) {
    const stripped = line.trim();
    if (stripped.startsWith('{') || stripped.startsWith('[')) {
      inJson = true;
      continue;
    }
    if (inJson) {
      if (stripped.endsWith('}') || stripped.endsWith(']')) {
        inJson = false;
      }
      continue;
    }

    const lowerLine = stripped.toLowerCase();
    if (
      [
        'here is a rewritten',
        "here's a rewritten",
        'here is the summary',
        'here is the executive summary',
        "here's the summary",
        "here's the executive summary",
        'here is a summary',
        'here is the json representation',
        "here's a summary",
        "here's the json representation",
        'rewritten executive summary',
      ].some((prefix) => lowerLine.includes(prefix))
    ) {
      continue;
    }

    cleanedLines.push(line);
  }

  summary = cleanedLines.join('\n').trim();

  // Remove any leading conversational phrases ending with a colon.
  summary = summary.replace(
    /^(here is|here's|this is|please find|summary of|executive summary of|rewritten executive summary|here's a rewritten|here is a rewritten)[\s\S]*?:\s*/i,
    ''
  );

  return summary.trim();
}

function hasUngroundedDollarOrDayFigures(summary: string, anchorsText: string): boolean {
  const dollarPattern = /\$\s?[\d,]+(?:\.\d+)?/g;
  const dayPattern = /\b\d+\s?-?days?\b|\b\w+\s?\(\d+\)\s?-?days?\b/gi;

  const dollars = summary.match(dollarPattern) || [];
  const days = summary.match(dayPattern) || [];

  const anchorsLower = anchorsText.toLowerCase();
  const summaryLower = summary.toLowerCase();

  for (const dollar of dollars) {
    const dollarNorm = dollar.toLowerCase().trim();
    if (!anchorsLower.includes(dollarNorm)) {
      let pos = summaryLower.indexOf(dollarNorm);
      while (pos !== -1) {
        const context = summaryLower.substring(pos, pos + 60);
        if (!context.includes('illustrative placeholder')) {
          return true;
        }
        pos = summaryLower.indexOf(dollarNorm, pos + 1);
      }
    }
  }

  for (const day of days) {
    const dayNorm = day.toLowerCase().trim();
    const digits = dayNorm.match(/\d+/);
    if (digits) {
      const digit = digits[0];
      if (!anchorsLower.includes(digit)) {
        let pos = summaryLower.indexOf(dayNorm);
        while (pos !== -1) {
          const context = summaryLower.substring(pos, pos + 60);
          if (!context.includes('illustrative placeholder')) {
            return true;
          }
          pos = summaryLower.indexOf(dayNorm, pos + 1);
        }
      }
    }
  }

  return false;
}

export function validateSummary(summary: string, anchorsText: string = ''): boolean {
  if (!summary) return false;
  if (summary.length < 200) return false;
  if (summary.startsWith('{') || summary.startsWith('[')) return false;
  if (summary.includes('```')) return false;

  const jsonMatches = summary.match(JSON_KEY_VALUE_PATTERN) || [];
  if (jsonMatches.length >= 2) return false;

  const lower = summary.toLowerCase();
  if (
    [
      'here is the json representation',
      "here's the json representation",
      'rewritten executive summary',
      'as an ai',
      'as a language model',
    ].some((phrase) => lower.includes(phrase))
  ) {
    return false;
  }

  if (anchorsText && hasUngroundedDollarOrDayFigures(summary, anchorsText)) {
    return false;
  }

  return true;
}

export class ReportWriter extends BaseAgent {
  readonly agentName = 'report_writer';
  protected maxTokens = 2048;

  async execute(state: ContractReviewState): Promise<Partial<ContractReviewState>> {
    const scores = state.risk_scores || [];
    const compliance = state.compliance_results || [];
    const edits = state.redline_edits || [];
    const clauses = state.clauses || [];
    const representedParty = state.represented_party || 'Client';

    // Top 3 highest-risk average calculation
    let overallScore = 0;
    if (scores.length > 0) {
      const sortedScores = scores.map((s) => s.score).sort((a, b) => b - a);
      const top3 = sortedScores.slice(0, 3);
      const sum = top3.reduce((acc, val) => acc + val, 0);
      overallScore = Math.round(sum / top3.length);
    }

    // Programmatic high-risk metrics (severity "high" or "critical")
    const highSeverityClauses = scores.filter((s) => s.severity === 'high' || s.severity === 'critical');
    const highRiskCount = highSeverityClauses.length;

    const clauseMap = new Map<number, any>();
    for (const c of clauses) {
      clauseMap.set(c.id, c);
    }

    const highRiskClausesInfo: string[] = [];
    for (const s of highSeverityClauses) {
      const c = clauseMap.get(s.clause_id);
      const heading = c ? c.heading : 'Unknown Heading';
      highRiskClausesInfo.push(
        `- Clause ${s.clause_id} (${heading}): Severity=${s.severity.toUpperCase()}, Score=${s.score}`
      );
    }
    const highRiskClausesText = highRiskClausesInfo.join('\n') || 'None';

    // Explicitly distinguish compliance violation counts (clauses vs. issues)
    const complianceViolationClauses =
      state.compliance_violation_clauses !== undefined
        ? state.compliance_violation_clauses
        : compliance.filter((r) => !r.compliant).length;

    const complianceViolationIssues =
      state.compliance_violation_issues !== undefined
        ? state.compliance_violation_issues
        : compliance.reduce((acc, r) => acc + (r.violations?.length || 0), 0);

    // Extract potential grounding anchors from clauses text to prevent hallucinated numbers
    const anchors: string[] = [];
    for (const c of clauses) {
      const textLower = (c.text || '').toLowerCase();
      if (textLower.includes('limit') || textLower.includes('cap')) {
        const moneyMatches = c.text.match(/\$\s?[\d,]+(?:\.\d+)?/g);
        if (moneyMatches) {
          anchors.push(`Clause ${c.id} ('${c.heading}') mentions liability limit/cap: ${moneyMatches.join(', ')}`);
        }
      }
      if (textLower.includes('cure') || textLower.includes('remedy') || textLower.includes('terminate')) {
        const dayMatches = c.text.match(/\b\d+\s?days?\b|\b\w+\s?\(\d+\)\s?days?\b/gi);
        if (dayMatches) {
          anchors.push(`Clause ${c.id} ('${c.heading}') mentions notice/cure period: ${dayMatches.join(', ')}`);
        }
      }
    }
    const anchorsText = anchors.length > 0 ? anchors.map((a) => `- ${a}`).join('\n') : 'None found in source text.';

    // Aggregate flags (exclude low risk/favorable flags to avoid executive summary mismatch)
    const flagCounts: Record<string, number> = {};
    for (const s of scores) {
      if (s.score >= 30 && s.flags) {
        for (const f of s.flags) {
          flagCounts[f] = (flagCounts[f] || 0) + 1;
        }
      }
    }
    const topFlags = Object.entries(flagCounts)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 5)
      .map(([f]) => f);

    const topCompliance: string[] = [];
    const seenCompliance = new Set<string>();
    for (const c of compliance) {
      if (!c.compliant && c.violations) {
        for (const v of c.violations) {
          const item = `${v.framework}: ${v.description}`;
          if (!seenCompliance.has(item)) {
            seenCompliance.add(item);
            topCompliance.push(item);
          }
        }
      }
    }
    const topComplianceList = topCompliance.slice(0, 5);

    let summaryPrompt = SUMMARY_PROMPT.replace(/{represented_party}/g, representedParty)
      .replace('{contract_name}', state.contract_name || 'Unknown')
      .replace('{clause_count}', String(clauses.length))
      .replace('{high_risk_count}', String(highRiskCount))
      .replace('{high_risk_clauses_list}', highRiskClausesText)
      .replace('{compliance_violation_clauses}', String(complianceViolationClauses))
      .replace('{compliance_violation_issues}', String(complianceViolationIssues))
      .replace('{overall_score}', String(overallScore))
      .replace('{top_flags}', topFlags.map((f) => `- ${f}`).join('\n') || 'None')
      .replace('{top_compliance}', topComplianceList.map((c) => `- ${c}`).join('\n') || 'None')
      .replace('{grounding_anchors}', anchorsText);

    let executiveSummary = '';
    for (let attempt = 1; attempt <= 3; attempt++) {
      const rawSummary = await this.callLlm(summaryPrompt);
      const cleaned = cleanSummaryText(rawSummary);
      if (validateSummary(cleaned, anchorsText)) {
        executiveSummary = cleaned;
        break;
      } else {
        console.warn(`[summary_validation_failed] Attempt: ${attempt}, Length: ${rawSummary.length}`);
        summaryPrompt = SUMMARY_PROMPT.replace(/{represented_party}/g, representedParty)
          .replace('{contract_name}', state.contract_name || 'Unknown')
          .replace('{clause_count}', String(clauses.length))
          .replace('{high_risk_count}', String(highRiskCount))
          .replace('{high_risk_clauses_list}', highRiskClausesText)
          .replace('{compliance_violation_clauses}', String(complianceViolationClauses))
          .replace('{compliance_violation_issues}', String(complianceViolationIssues))
          .replace('{overall_score}', String(overallScore))
          .replace('{top_flags}', topFlags.map((f) => `- ${f}`).join('\n') || 'None')
          .replace('{top_compliance}', topComplianceList.map((c) => `- ${c}`).join('\n') || 'None')
          .replace('{grounding_anchors}', anchorsText) +
          '\n\nCRITICAL: Your previous response contained conversational comments, raw JSON, markdown code blocks, or was too short. DO NOT output JSON or any conversational preambles like \'Here is the summary\'. Write ONLY the clean, user-facing executive summary paragraphs directly. Plain text only.';
      }
    }

    if (!executiveSummary) {
      const rawSummary = await this.callLlm(summaryPrompt);
      executiveSummary = cleanSummaryText(rawSummary) || 'Failed to generate executive summary.';
    }

    // Prepend perspective statement
    const prefix = `This analysis is prepared from the ${representedParty}'s perspective.\n\n`;
    if (!executiveSummary.startsWith('This analysis is prepared')) {
      executiveSummary = prefix + executiveSummary;
    }

    const extractionWarning = state.extraction_integrity_warning;
    if (extractionWarning) {
      console.error(`[report_includes_extraction_integrity_warning] Warning: ${extractionWarning}`);
    }

    const report: FinalReport = {
      contract_id: state.contract_id,
      contract_name: state.contract_name || 'Unknown',
      overall_score: overallScore,
      clause_count: clauses.length,
      high_risk_count: highRiskCount,
      violation_count: complianceViolationIssues,
      compliance_violation_clauses: complianceViolationClauses,
      compliance_violation_issues: complianceViolationIssues,
      executive_summary: executiveSummary,
      risk_scores: scores,
      compliance_results: compliance,
      redline_edits: edits,
      represented_party: representedParty,
      extraction_integrity_warning: extractionWarning || null,
      generated_at: new Date(),
    };

    return {
      report,
      pipeline_complete: true,
    };
  }
}
